src/awkward/operations/ak_from_iter.py

```diff
@@ -98,14 +109,16 @@ def _impl(iterable, highlevel, behavior, allow_record, initial, resize):
     formstr, length, buffers = builder.to_buffers()
     form = ak.forms.from_json(formstr)
 
-    return ak.operations.ak_from_buffers._impl(
+    outer = ak.operations.ak_from_buffers._impl(
         form,
         length,
         buffers,
         buffer_key="{form_key}-{attribute}",
         backend="cpu",
         byteorder=ak._util.native_byteorder,
-        highlevel=highlevel,
+        highlevel=False,
         behavior=behavior,
         simplify=True,
-    )[0]
+    )
+
+    return wrap_layout(outer[0], highlevel=highlevel, behavior=behavior)
```

Comment:
It *does* change behaviour, and I originally documented it in another pr. Oops!

If we don't do this, then behaviour classes with custom get item overloads can return non layout types. In my view the way we're using getitem here is an implementation detail rather than something users should play around with.

The specific example of what breaks without this change is when a string is passed into `from_iter` - in main, this returns a python string. With this PR, it returns an array of characters (strings are iterables of char).

I think that's a good change, though it is be breaking if users want a python string to round trip. It's your call whether this is a bug fix!

Reply:
If a string/bytestring is passed directly into `ak.from_iter` or `ak.Array` without being wrapped in a length-1 list, it wouldn't be too surprising to see it turned into an array of characters. But what about when it's being picked up implicitly as in the examples in https://github.com/scikit-hep/awkward/pull/2353#discussion_r1156261075?

Replace the string/bytestring with a scalar value like `3`, and you'd expect it to broadcast as though it were `[3]`. A string in a broadcasting situation should behave like an atom, like `3`, not as a list of characters that would have to have the same length as the array it's being broadcasted against.

```python
all_my_data["new field"] = "label"
```

looks like it ought to duplicate `"label"` and attach it to `all_my_data` as a `new field`. If this tries to broadcast it instead, it fail if the length of `all_my_data` is not equal to the length of `"label"`, a confusing situation whether it works or doesn't.

Is this going to happen, considering that `"label"` has to be turned into an Awkward Array before the broadcasting can take place, and that conversion probably uses `ak.from_iter`?

Reply:
It does, but it also does so in `main`.

```python
>>> import awkward as ak
>>> all_my_data = ak.Array([{"x": 1}, {"x": 2}, {"x": 3}, {"x": 4}, {"x": 5}])
>>> all_my_data["new field"] = "label"
<Array [{x: 1, 'new field': 108}, ..., {...}] type='5 * {x: int64, "new fie...'>
>>> all_my_data.show()
[{x: 1, 'new field': 108},
 {x: 2, 'new field': 97},
 {x: 3, 'new field': 98},
 {x: 4, 'new field': 101},
 {x: 5, 'new field': 108}]
```

It didn't in Awkward 1.10.2.

```python
>>> import awkward as ak
>>> all_my_data = ak.Array([{"x": 1}, {"x": 2}, {"x": 3}, {"x": 4}, {"x": 5}])
>>> all_my_data["new field"] = "label"
>>> all_my_data.tolist()
[
    {'x': 1, 'new field': 'label'},
    {'x': 2, 'new field': 'label'},
    {'x': 3, 'new field': 'label'},
    {'x': 4, 'new field': 'label'},
    {'x': 5, 'new field': 'label'},
]
```

Do you agree that this is a bug? I can open an issue, but I don't know if this PR is going to be a step toward or away from that.